### PR TITLE
gh-122681: merge m_atan2() and c_atan2() helper functions

### PR DESCRIPTION
**Pull request by skirpichev**
* Issue: gh-122681

---------------------

For reviewers: as it mentioned in the issue, I think it worth to try removing these workarounds.  Most cmath's functions are using libm's atan2() and this don't poses any problems.  If that does make sense, here is a draft pr to test this: https://github.com/python/cpython/pull/122715.

### CODE REVIEW

File: Tools/requirements-hypothesis.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # Requirements file for hypothesis that
 # we use to run our property-based tests in CI.
 
+attrs<=23.2.0
```
Comment (vstinner) [inline Tools/requirements-hypothesis.txt]: unrelated change

File: Modules/_math.h [hunk 2]
```diff
@@ -23,3 +23,42 @@ _Py_log1p(double x)
...
 
 #define m_log1p _Py_log1p
+
+/*
+   wrapper for atan2 that deals directly with special cases before
+   delegating to the platform libm for the remaining cases.  This
+   is necessary to get consistent behaviour across platforms.
+   Windows, FreeBSD and alpha Tru64 are amongst platforms that don't
+   always follow C99.  Windows screws up atan2 for inf and nan, and
+   alpha Tru64 5.1 doesn't follow C99 for atan2(0., 0.).
```
Comment (vstinner) [inline Modules/_math.h:33]: I'm curious if Windows and FreeBSD still need these workaround in 2024. But I don't ask you to test ;-)

File: Misc/NEWS.d/next/Library/2024-08-05-10-52-51.gh-issue-122681.q0T58t.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Merge ``m_atan2()`` and ``c_atan2()`` helper function, move to
+``Modules/_math.h``.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2024-08-05-10-52-51.gh-issue-122681.q0T58t.rst]: I don't think that this change benefits to users, you can remove the NEWS entry.

File: Tools/requirements-hypothesis.txt (hunk 1 above)
Comment (skirpichev) [reply Tools/requirements-hypothesis.txt]: was a quick test for https://github.com/python/cpython/pull/122687, sorry; reverted

File: Misc/NEWS.d/next/Library/2024-08-05-10-52-51.gh-issue-122681.q0T58t.rst (hunk 3 above)
Comment (skirpichev) [reply Misc/NEWS.d/next/Library/2024-08-05-10-52-51.gh-issue-122681.q0T58t.rst]: Done.  Could you, please, then add "skip news" label?

File: Modules/_math.h (hunk 2 above)
Comment (skirpichev) [reply Modules/_math.h:33]: > I'm curious if Windows and FreeBSD still need these workaround in 2024.

I doubt, especially FreeBSD.

As I said in the description, most functions in the cmath works with libm's atan2 function (e.g. acos).  I think it's safe to remove these helpers.  But this should be heavily tested.  Here is a draft pr with this attempt: https://github.com/python/cpython/pull/122715

Comment (vstinner) [reply Modules/_math.h:33]: > I doubt, especially FreeBSD.

Right. When I added math.fma(), I found a bug on FreeBSD. It takes a while to get it fixed upstream, but it's being fixed: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=277783 :-)

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
I merged the main branch into this PR to retrieve the Hypothesis workaround (pin attrs version).

**Comment by vstinner** [comment]
Merged. Thank you.
